src/Features/Protocol/LanguageServices/RoslynLanguageService.cs

```diff
@@ -0,0 +1,673 @@
+        /// https://microsoft.github.io/language-server-protocol/specification#textDocument_definition
+        /// </summary>
+        /// <param name="solution">the solution containing the request.</param>
+        /// <param name="request">the document position of the symbol to go to.</param>
+        /// <param name="cancellationToken">a cancellation token.</param>
+        /// <returns>the location of a given symbol.</returns>
+        public Task<VSLocation[]> GoToDefinitionAsync(Solution solution, TextDocumentPositionParams request, CancellationToken cancellationToken)
+        {
+            return GetDefinitionAsync(solution, request, typeOnly: false, cancellationToken);
+        }
+
+        /// <summary>
+        /// Answers a goto type definition request by returning the location of a given type definition.
+        /// https://microsoft.github.io/language-server-protocol/specification#textDocument_typeDefinition
+        /// </summary>
+        /// <param name="solution">the solution containing the request.</param>
+        /// <param name="request">the document position of the type to go to.</param>
+        /// <param name="cancellationToken">a cancellation token.</param>
+        /// <returns>the location of a type definition.</returns>
+        public Task<VSLocation[]> GoToTypeDefinitionAsync(Solution solution, TextDocumentPositionParams request, CancellationToken cancellationToken)
+        {
+            return GetDefinitionAsync(solution, request, typeOnly: true, cancellationToken);
+        }
+
+        /// <summary>
+        /// Answers a find references request by returning the location of references to the given symbol.
+        /// https://microsoft.github.io/language-server-protocol/specification#textDocument_references
+        /// </summary>
+        /// <param name="solution">the solution containing the symbol.</param>
+        /// <param name="request">the request symbol document location.</param>
+        /// <param name="cancellationToken">a cancellation token.</param>
+        /// <returns>a list of locations of references to the given symbol.</returns>
+        public async Task<VSLocation[]> FindAllReferencesAsync(Solution solution, ReferenceParams request, CancellationToken cancellationToken)
+        {
+            var locations = ArrayBuilder<VSLocation>.GetInstance();
+
+            var document = solution.GetDocument(request.TextDocument.Uri);
+            if (document == null)
+            {
+                return locations.ToArrayAndFree();
+            }
+
+            var findUsagesService = document.Project.LanguageServices.GetService<IFindUsagesService>();
+            var position = await document.GetPositionAsync(request.Position.ToLinePosition(), cancellationToken).ConfigureAwait(false);
+
+            var context = new SimpleFindUsagesContext(cancellationToken);
+
+            await findUsagesService.FindReferencesAsync(document, position, context).ConfigureAwait(false);
```

Comment:
this likely is not what VSCode wants for FindAllRefs.  This would need to be designed given their system.  for example, if you did GoToDef on Foo.Dispose, this would find all the IDisposable.Dispose methods in the world.  That's ok in Roslyn where we bucket things by definition.  but that's not going on in VSCode, which means you're going to get a huge amount of cruft thrown at you you probably don't want/need.

Reply:
this is how live share currently works. and live share is actually wire it over internet where vscode is just named pipe line on same machine.

so, I dont believe things wasn't problem over internet becomes problem over pipe.

and this kind of perf concern can be addressed once we make functionality work.